neonvm/runner/main.go

```diff
@@ -61,6 +61,12 @@ const (
 	sshAuthorizedKeysDiskPath   = "/vm/images/ssh-authorized-keys.iso"
 	sshAuthorizedKeysMountPoint = "/vm/ssh"
 
+	// See #775 and its links.
+	// * cache.writeback=off - forces writes to be synchronous (avoid hidden caching)
```

Comment:
Having `cache.writeback=off` means qemu makes sure to flush the data to disk on write requests. My understanding is this is like calling fsync on every write, which would have a major performance impact. Since we don't care about persisting data on host we can go with `cache.writeback=on`.

Source: https://www.qemu.org/docs/master/system/qemu-manpage.html

Reply:
Ah right, good catch! Honestly I'm not sure whether `cache.writeback=off` should have any effect if `cache.direct` is used... Because `cache.direct=on` forces all writes to use direct IO (rather than via the page cache) and `cache.writeback=on` seems to say that it's more about returning early and doing flushes asynchronously.

Given that, I'm happy to set `cache.writeback=on` — on the off-chance that it does have _some_ effect even if `cache.direct=on`, I'd rather go with the one that is more efficient by default.

I'll also update this PR so that we can configure QEMU cache settings with an arg to `neonvm-controller` that propagates to `neonvm-runner` — that way, we can easily test this out on staging / pre-prod without needing to recompile each time.

Reply:
> Ah right, good catch! Honestly I'm not sure whether cache.writeback=off should have any effect if cache.direct is used... Because cache.direct=on forces all writes (and reads I geuss?) to use direct IO (rather than via the page cache) and cache.writeback=on seems to say that it's more about returning early and doing flushes asynchronously.

I think having `cache.direct=on` doesn't mean all writes would go to the disk, but it means we only use guest's memory and avoid host's page cache. Here is the definition from the QEMU manpage:

> The host page cache can be avoided with `cache.direct=on`. This will attempt to do disk IO directly to the guest’s memory. QEMU may still perform an internal copy of the data.

---
> I'll also update this PR so that we can configure QEMU cache settings with an arg to neonvm-controller that propagates to neonvm-runner — that way, we can easily test this out on staging / pre-prod without needing to recompile each time.

Sounds good.

Reply:
Ah ok, so what I can find is:

- `cache.direct=on` means setting `O_DIRECT` when opening the backing file
- `cache.writeback=off` means setting `O_DSYNC` when opening the backing file (i.e. every write flushes, like you mentioned)

Some of this info is collected from wiki pages of projects built on QEMU, but I _think_ these are the semantics.

So, given that, I'd say yeah we definitely should have `cache.writeback=off`

Reply:
When the disks are ephemeral, why do we want to call fsync on every write? It makes every write much slower.

Reply:
Oh sorry, typo. Meant to say `writeback=on` :sweat_smile: